pkg/operator/vsphereproblemdetector/vsphere_problem_detector_deployment.go

```diff
@@ -0,0 +1,189 @@
+	f = f.WithInformers(
+		clients.OperatorClient.Informer(),
+		clients.KubeInformers.InformersFor(csoclients.OperatorNamespace).Apps().V1().Deployments().Informer())
+
+	c := &VSphereProblemDetectorDeploymentController{
+		operatorClient: clients.OperatorClient,
+		kubeClient:     clients.KubeClient,
+		versionGetter:  versionGetter,
+		eventRecorder:  eventRecorder,
+		targetVersion:  targetVersion,
+		factory:        f,
+	}
+	return c
+}
+
+func (c *VSphereProblemDetectorDeploymentController) Sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	klog.V(4).Infof("VSphereProblemDetectorDeploymentController sync started")
+	defer klog.V(4).Infof("VSphereProblemDetectorDeploymentController sync finished")
+
+	opSpec, opStatus, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if opSpec.ManagementState != operatorapi.Managed {
+		return nil
+	}
+	required := c.getRequiredDeployment(opSpec)
+	lastGeneration := resourcemerge.ExpectedDeploymentGeneration(required, opStatus.Generations)
+	deployment, _, err := resourceapply.ApplyDeployment(c.kubeClient.AppsV1(), c.eventRecorder, required, lastGeneration)
+	if err != nil {
+		// This will set Degraded condition
+		return err
+	}
+
+	// Available: at least one replica is running
+	deploymentAvailable := operatorapi.OperatorCondition{
+		Type: deploymentControllerName + operatorapi.OperatorStatusTypeAvailable,
+	}
+	if deployment.Status.AvailableReplicas > 0 {
+		deploymentAvailable.Status = operatorapi.ConditionTrue
+	} else {
+		deploymentAvailable.Status = operatorapi.ConditionFalse
+		deploymentAvailable.Reason = "WaitDeployment"
+		deploymentAvailable.Message = "Waiting for a Deployment pod to start"
+	}
+
+	// Not progressing: all replicas are at the latest version && Deployment generation matches
+	deploymentProgressing := operatorapi.OperatorCondition{
+		Type: deploymentControllerName + operatorapi.OperatorStatusTypeProgressing,
+	}
+	if deployment.Status.ObservedGeneration != deployment.Generation {
+		deploymentProgressing.Status = operatorapi.ConditionTrue
+		deploymentProgressing.Reason = "NewGeneration"
+		msg := fmt.Sprintf("desired generation %d, current generation %d", deployment.Generation, deployment.Status.ObservedGeneration)
+		deploymentProgressing.Message = msg
+	} else {
+		if deployment.Spec.Replicas != nil {
+			if deployment.Status.UpdatedReplicas == *deployment.Spec.Replicas {
+				deploymentProgressing.Status = operatorapi.ConditionFalse
+				// All replicas were updated, set the version
+				c.versionGetter.SetVersion(deploymentControllerName, c.targetVersion)
+			} else {
+				msg := fmt.Sprintf("%d out of %d pods running", deployment.Status.UpdatedReplicas, *deployment.Spec.Replicas)
+				deploymentProgressing.Status = operatorapi.ConditionTrue
+				deploymentProgressing.Reason = "WaitDeployment"
+				deploymentProgressing.Message = msg
+			}
+		}
+	}
+
+	resourcemerge.SetDeploymentGeneration(&opStatus.Generations, deployment)
```

Comment:
Generation will be set below in `UpdateStatus`, so this update is useless. And yes, CSIDriverOperatorDeploymentController has it wrong too.